pkg/syncer/synctarget/tunneler_reconciler.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package synctarget
+
+import (
+	"context"
+	"sync"
+
+	utilserrors "k8s.io/apimachinery/pkg/util/errors"
+
+	workloadv1alpha1 "github.com/kcp-dev/kcp/sdk/apis/workload/v1alpha1"
+)
+
+type tunnelerReconciler struct {
+	tunnelersLock sync.RWMutex
+	tunnelers     map[workloadv1alpha1.TunnelWorkspace]tunneler
+
+	startShardTunneler func(ctx context.Context, shardURL workloadv1alpha1.TunnelWorkspace)
+}
+
+func (c *tunnelerReconciler) reconcile(ctx context.Context, syncTarget *workloadv1alpha1.SyncTarget) (reconcileStatus, error) {
+	requiredShards := map[workloadv1alpha1.TunnelWorkspace]bool{}
+	if syncTarget != nil {
+		for _, shardURL := range syncTarget.Status.TunnelWorkspaces {
+			requiredShards[shardURL] = true
+		}
+	}
+
+	c.tunnelersLock.Lock()
+	defer c.tunnelersLock.Unlock()
+
+	// Remove obsolete tunnelers that don't have a shard anymore
+	for shardURL, shardTunneler := range c.tunnelers {
+		if _, ok := requiredShards[shardURL]; ok {
+			// The tunnelers are still expected => don't remove them
+			continue
+		}
+		// The tunnelers should not be running
+		// Stop them and remove it from the list of started shard tunneler
+		shardTunneler.stop()
+		delete(c.tunnelers, shardURL)
+	}
+
+	var errs []error
+	// Create and start missing tunnelers
+	for shardURL := range requiredShards {
+		if _, ok := c.tunnelers[shardURL]; ok {
+			// The tunnelers are already started
+			continue
+		}
+
+		// Start the tunnelers
+		shardTunnelerContext, cancelFunc := context.WithCancel(ctx)
+
+		// Create the tunneler
+		c.startShardTunneler(shardTunnelerContext, shardURL)
+		c.tunnelers[shardURL] = tunneler{
+			cancelFunc,
+		}
+	}
+	return reconcileStatusContinue, utilserrors.NewAggregate(errs)
+}
+
+type tunneler struct {
```

Comment:
Why do we need to allocate a new struct just to hold onto the function pointer?

Reply:
Just a rest of copied code. I'll clean this up.

Reply:
Fixed